File: go-backend-interview-prep/02_data_structures/07_pointers_intro/main.py
# Python equivalent of Pointers (compare with main.go)
# Python has no explicit pointers.
# Mutable objects (list, dict, class instances) use reference semantics.
# Immutable objects (int, str, tuple) behave like pass-by-value.


class Config:
    def __init__(self, host: str = "", port: int = 0, verbose: bool = False):
        self.host = host
        self.port = port
        self.verbose = verbose

    def __repr__(self):
        return f"Config(host={self.host!r}, port={self.port}, verbose={self.verbose})"


def increment_immutable(x: int) -> int:
    """Cannot modify caller's int. Must return new value."""
    return x + 1


def enable_verbose(c: Config) -> None:
    """Mutable object: modifies the original (like Go pointer)."""
    c.verbose = True


def no_change(n: int) -> None:
    n = 999  # rebinds local name; original unchanged


def create_config() -> Config:
    """Returning a new object (Python manages memory automatically)."""
    return Config(host="api.example.com", port=443)


def main():
    # --- No explicit & and * in Python ---
    # Python uses id() to see "address"
    x = 42
    print("x:", x)
    print("id(x):", id(x))

    # --- Mutating requires reassignment for immutables ---
    x = increment_immutable(x)
    print("x after increment:", x)

    # --- Mutable objects behave like pointers ---
    print("\n--- Mutable object (like pointer) ---")
    cfg = Config(host="localhost", port=8080)
    print("cfg.host:", cfg.host)
    cfg.port = 9090
    print("cfg.port:", cfg.port)

    # --- Pass mutable object to function ---
    print("\n--- Pass mutable (like *Config) ---")
    enable_verbose(cfg)
    print(f"after enable_verbose: {cfg}")

    # --- Immutable (like Go pass-by-value) ---
    print("\n--- Immutable int (like pass-by-value) ---")
    num = 10
    no_change(num)
    print("num unchanged:", num)

    # --- None = Go's nil ---
    print("\n--- None (like nil) ---")
    ref = None
    print("ref:", ref)
    print("ref is None:", ref is None)
    # Accessing ref.something would raise AttributeError (like nil deref panic)

    # --- Object identity ---
    print("\n--- Identity ---")
    a = [1, 2, 3]
    b = [1, 2, 3]
    print("a == b:", a == b)     # True (same value)
    print("a is b:", a is b)     # False (different objects)

    # --- Factory function ---
    print("\n--- Factory ---")
    cp = create_config()
    print(f"created: {cp}")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/02_data_structures/07_pointers_intro/main.go
package main

import "fmt"

type Config struct {
	Host    string
	Port    int
	Verbose bool
}

func main() {
	// --- Basics: & and * ---
	x := 42
	p := &x // p is *int, holds address of x

	fmt.Println("x:", x)
	fmt.Println("&x (address):", p)
	fmt.Println("*p (deref):", *p)

	// --- Modify through pointer ---
	*p = 100
	fmt.Println("x after *p = 100:", x) // 100

	// --- Pointer to a struct ---
	fmt.Println("\n--- Struct pointer ---")
	cfg := &Config{Host: "localhost", Port: 8080}
	fmt.Println("cfg.Host:", cfg.Host) // auto-deref (no ->)
	cfg.Port = 9090
	fmt.Println("cfg.Port:", cfg.Port)

	// --- Pass by pointer to mutate ---
	fmt.Println("\n--- Pass by pointer ---")
	val := 10
	fmt.Println("before:", val)
	increment(&val)
	fmt.Println("after increment:", val) // 11

	// --- Pass struct by pointer ---
	fmt.Println("\n--- Struct by pointer ---")
	user := Config{Host: "example.com", Port: 80}
	enableVerbose(&user)
	fmt.Printf("after enableVerbose: %+v\n", user)

	// --- Pass by value (does NOT mutate) ---
	fmt.Println("\n--- Pass by value (no mutation) ---")
	num := 10
	noChange(num)
	fmt.Println("num unchanged:", num)

	// --- new() allocates zero-value and returns pointer ---
	fmt.Println("\n--- new() ---")
	np := new(int)
	fmt.Println("*new(int):", *np) // 0
	*np = 55
	fmt.Println("after assignment:", *np)

	// --- Nil pointer ---
	fmt.Println("\n--- Nil pointer ---")
	var nilPtr *int
	fmt.Println("nilPtr:", nilPtr)             // <nil>
	fmt.Println("nilPtr == nil:", nilPtr == nil) // true
	// *nilPtr would PANIC: "invalid memory address"

	// --- Pointer comparison ---
	fmt.Println("\n--- Pointer comparison ---")
	a := 10
	b := 10
	pa := &a
	pb := &b
	fmt.Println("pa == pb:", pa == pb) // false (different addresses)
	fmt.Println("*pa == *pb:", *pa == *pb) // true (same value)

	// --- Local variable pointer is safe (escape analysis) ---
	fmt.Println("\n--- Escape analysis ---")
	cp := createConfig()
	fmt.Printf("returned pointer: %+v\n", *cp)
}

func increment(p *int) {
	*p++ // modifies the original
}

func enableVerbose(c *Config) {
	c.Verbose = true // modifies the original struct
}

func noChange(n int) {
	n = 999 // modifies local copy only
}

// Returning a pointer to a local variable is safe in Go.
// The compiler moves it to the heap via escape analysis.
func createConfig() *Config {
	c := Config{Host: "api.example.com", Port: 443}
	return &c
}
